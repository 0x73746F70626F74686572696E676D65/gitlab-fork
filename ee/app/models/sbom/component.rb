# frozen_string_literal: true

module Sbom
  class Component < ApplicationRecord
    enum component_type: {
      library: 0
    }

    validates :component_type, presence: true
    validates :name, presence: true, length: { maximum: 255 }
  end
end
