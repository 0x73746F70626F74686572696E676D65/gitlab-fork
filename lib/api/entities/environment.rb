# frozen_string_literal: true

module API
  module Entities
    class Environment < Entities::EnvironmentBasic
      include RequestAwareEntity
      include Gitlab::Utils::StrongMemoize

      expose :tier
      expose :project, using: Entities::BasicProjectDetails
      expose :last_deployment, using: Entities::Deployment, if: { last_deployment: true }
      expose :state

      private

      alias_method :environment, :object

      def can_read_pod_logs?
        strong_memoize(:can_read_pod_logs) do
          current_user&.can?(:read_pod_logs, environment.project)
        end
      end

      def cluster
        strong_memoize(:cluster) do
          environment&.last_deployment&.cluster
        end
      end

      def current_user
        options[:current_user]
      end
    end
  end
end
