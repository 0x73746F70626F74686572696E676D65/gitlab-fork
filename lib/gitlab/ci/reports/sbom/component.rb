# frozen_string_literal: true

module Gitlab
  module Ci
    module Reports
      module Sbom
        class Component
          attr_reader :component_type, :name, :version

          def initialize(component = {})
            @component_type = component['type']
            @name = component['name']
            @version = component['version']
          end
        end
      end
    end
  end
end
