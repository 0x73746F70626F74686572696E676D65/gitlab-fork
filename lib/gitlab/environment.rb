# frozen_string_literal: true

module Gitlab
  module Environment
    def self.hostname
      @hostname ||= ENV['HOSTNAME'] || Socket.gethostname
    end

    def self.qa_user_agent
      ENV['GITLAB_QA_USER_AGENT']
    end
  end
end
