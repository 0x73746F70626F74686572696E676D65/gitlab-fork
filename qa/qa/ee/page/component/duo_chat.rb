# frozen_string_literal: true

module QA
  module EE
    module Page
      module Component
        class DuoChat < QA::Page::Base
          view 'ee/app/assets/javascripts/ai/tanuki_bot/components/app.vue' do
            # components are derived from gitlab/ui
          end

          def send_duo_chat_prompt(prompt)
            fill_element('chat-prompt-input', prompt)
            click_element('paper-airplane-icon')
            wait_for_requests
          end

          def clear_chat_history
            send_duo_chat_prompt('/clean')
          end

          def empty_state?
            has_element?('gl-empty-state-content')
          end

          def has_response?(expected_response)
            has_css?('.duo-chat-message p', text: expected_response, wait: 30)
          end

          def number_of_messages
            find_all('.duo-chat-message').size
          end
        end
      end
    end
  end
end
