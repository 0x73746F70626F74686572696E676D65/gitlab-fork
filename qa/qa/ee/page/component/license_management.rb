# frozen_string_literal: true

module QA
  module EE
    module Page
      module Component
        module LicenseManagement
          extend QA::Page::PageConcern

          def self.prepended(base)
            super

            base.class_eval do
              view 'app/assets/javascripts/reports/components/report_item.vue' do
                element :report_item_row
              end

              view 'app/assets/javascripts/reports/components/issue_status_icon.vue' do
                element :icon_status, ':data-qa-selector="`status_${status}_icon`" ' # rubocop:disable QA/ElementWithPattern
              end

              view 'ee/app/assets/javascripts/vue_shared/license_compliance/mr_widget_license_report.vue' do
                element :license_report_widget
                element :manage_licenses_button
              end

              view 'app/assets/javascripts/vue_merge_request_widget/components/extensions/base.vue' do
                element :mr_widget_extension
              end

              view 'app/assets/javascripts/vue_merge_request_widget/components/action_buttons.vue' do
                element :mr_widget_extension_actions_button
              end

              view 'app/assets/javascripts/vue_merge_request_widget/components/extensions/child_content.vue' do
                element :child_content
              end
            end
          end

          def has_approved_license?(name)
            content_element = feature_flag_controlled_element(:refactor_license_compliance_extension,
                                            :child_content,
                                            :report_item_row)
            within_element(content_element, text: name) do
              has_element?(:status_success_icon, wait: 1)
            end
          end

          def has_denied_license?(name)
            content_element = feature_flag_controlled_element(:refactor_license_compliance_extension,
                                                              :child_content,
                                                              :report_item_row)
            within_element(content_element, text: name) do
              has_element?(:status_failed_icon, wait: 1)
            end
          end

          def click_manage_licenses_button
            previous_page = page.current_url

            widget_element = feature_flag_controlled_element(:refactor_license_compliance_extension,
                                                             :mr_widget_extension,
                                                             :license_report_widget)
            within_element(widget_element) do
              if widget_element == :mr_widget_extension
                click_element(:mr_widget_extension_actions_button, text: 'Manage Licenses')
              else
                click_element(:manage_licenses_button)
              end
            end
            # TODO workaround for switched to a new window UI
            wait_until(max_duration: 15, reload: false) do
              page.current_url != previous_page
            end
          end
        end
      end
    end
  end
end
