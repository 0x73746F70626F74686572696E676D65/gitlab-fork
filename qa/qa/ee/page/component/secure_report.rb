# frozen_string_literal: true

module QA
  module EE
    module Page
      module Component
        module SecureReport
          extend QA::Page::PageConcern

          def self.prepended(base)
            super

            base.class_eval do
              view 'ee/app/assets/javascripts/security_dashboard/components/filters/simple_filter.vue' do
                element :filter_dropdown, ':data-qa-selector="qaSelector"' # rubocop:disable QA/ElementWithPattern
                element :filter_dropdown_content
              end

              view 'ee/app/assets/javascripts/security_dashboard/components/security_dashboard_table.vue' do
                element :security_report_content
              end

              view 'ee/app/assets/javascripts/security_dashboard/components/shared/filters/activity_filter.vue' do
                element :filter_activity_dropdown
              end

              view 'ee/app/assets/javascripts/security_dashboard/components/shared/filters/status_filter.vue' do
                element :filter_status_dropdown
              end

              view 'ee/app/assets/javascripts/security_dashboard/components/
                    shared/vulnerability_report/vulnerability_list.vue' do
                element :vulnerability_status_content
              end
            end
          end

          def filter_report_type(report, project_filter = true)
            click_element(:filter_tool_dropdown)

            if project_filter
              find("[data-testid='listbox-item-GitLab.#{report.upcase.tr(" ", "_")}']").click
            else
              find("[data-testid='listbox-item-#{report.upcase.tr(" ", "_")}']").click
            end
            # Click the dropdown to close the modal and ensure it isn't open if this function is called again
            click_element(:filter_tool_dropdown)
          end

          def filter_by_status(statuses)
            wait_until(max_duration: 30, message: "Waiting for status dropdown element to appear") do
              has_element?(:filter_status_dropdown)
            end

            # Retry on exception to avoid ElementNotFound errors when clicks are sent too fast for the UI to update
            retry_on_exception(sleep_interval: 2, message: "Retrying status click until current url matches state") do
              find(status_dropdown_button_selector, wait: 5).click
              find(status_item_selector('ALL')).click
              statuses_list(statuses).each do |status|
                find(status_item_selector(status)).click
                wait_for_requests # It takes a moment to update the page after changing selections
              end
              find(status_dropdown_button_selector, wait: 5).click
              state = statuses_list(statuses).map { |item| "state=#{item}" }.join("&")
              page.current_url.downcase.include?(state)
            end
          end

          def statuses_list(statuses)
            statuses.map do |status|
              case status
              when /all/i
                'all'
              when /needs triage/i
                'detected'
              else
                status
              end
            end
          end

          def status_dropdown_button_selector
            "[data-testid='filter-status-dropdown'] > button"
          end

          def status_item_selector(status)
            "[data-testid='listbox-item-#{status.upcase}']"
          end

          def filter_by_activity(activity_name)
            find(activity_dropdown_button_selector, wait: 5).click
            find(activity_item_selector(activity_name)).click
          end

          def activity_dropdown_button_selector
            "[data-testid='filter-activity-dropdown'] > button"
          end

          def activity_item_selector(activity_name)
            "[data-testid='listbox-item-#{activity_name}']"
          end

          def has_vulnerability?(name)
            retry_until(reload: true, sleep_interval: 10, max_attempts: 6, message: "Retry for vulnerability text") do
              has_element?(:vulnerability, text: name)
            end
          end

          def has_vulnerability_info_content?(name)
            retry_until(reload: true, sleep_interval: 2, max_attempts: 3) do
              click_link('Security') unless has_element?(:security_report_content)
              has_element?(:vulnerability_info_content, text: name)
            end
          end

          def has_status?(status, vulnerability_name)
            retry_until(reload: true, sleep_interval: 3, raise_on_failure: false) do
              # Capitalizing first letter in each word to account for "Needs Triage" state
              has_element?(
                :vulnerability_status_content,
                status_description: vulnerability_name,
                text: "#{status.split.map(&:capitalize).join(' ')}"
              )
            end
          end
        end
      end
    end
  end
end
