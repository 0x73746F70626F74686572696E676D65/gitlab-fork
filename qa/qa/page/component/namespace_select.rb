# frozen_string_literal: true

module QA
  module Page
    module Component
      module NamespaceSelect
        extend QA::Page::PageConcern

        def self.included(base)
          super

          base.view "app/assets/javascripts/groups_projects/components/transfer_locations.vue" do
            element :namespaces_list
            element :namespaces_list_groups
            element :namespaces_list_item
            element :namespaces_list_search
          end
        end

        def select_namespace(item)
          click_element :namespaces_list

          within_element(:namespaces_list) do
            fill_element(:namespaces_list_search, item)

            wait_for_requests

            click_element(:namespaces_list_item, text: item)
          end
        end
      end
    end
  end
end
