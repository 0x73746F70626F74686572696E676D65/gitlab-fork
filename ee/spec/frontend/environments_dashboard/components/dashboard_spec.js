import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import { GlButton } from '@gitlab/ui';
import ProjectSelector from '~/vue_shared/components/project_selector/project_selector.vue';
import component from 'ee/environments_dashboard/components/dashboard/dashboard.vue';
import ProjectHeader from 'ee/environments_dashboard/components/dashboard/project_header.vue';
import Environment from 'ee/environments_dashboard/components/dashboard/environment.vue';

import environment from './mock_environment.json';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('dashboard', () => {
  const Component = localVue.extend(component);
  let wrapper;
  let propsData;

  beforeEach(() => {
    propsData = {
      addPath: 'mock-addPath',
      listPath: 'mock-listPath',
      emptyDashboardSvgPath: '/assets/illustrations/operations-dashboard_empty.svg',
      emptyDashboardHelpPath: '/help/user/operations_dashboard/index.html',
    };

    wrapper = mount(Component, {
      propsData,
      localVue,
      methods: {
        fetchProjects: () => {},
      },
      sync: false,
    });
  });

  it('should match the snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the dashboard title', () => {
    expect(wrapper.find('.js-dashboard-title').text()).toBe('Environments Dashboard');
  });

  describe('add projects button', () => {
    let button;

    beforeEach(() => {
      button = wrapper.find(GlButton);
    });

    it('is labelled correctly', () => {
      expect(button.text()).toBe('Add projects');
    });

    it('should show the modal on click', done => {
      button.trigger('click');
      wrapper.vm.$nextTick(() => {
        expect(wrapper.find(ProjectSelector)).toExist();
        done();
      });
    });
  });

  describe('wrapped components', () => {
    beforeEach(done => {
      wrapper.vm.projects = [
        {
          id: 0,
          name: 'test',
          namespace: { name: 'test', id: 0 },
          environments: [{ ...environment, id: 0 }, environment],
        },
        { id: 1, name: 'test', namespace: { name: 'test', id: 0 }, environments: [environment] },
      ];
      wrapper.vm.$nextTick(() => done());
    });

    describe('project header', () => {
      it('should have one project header per project', () => {
        const headers = wrapper.findAll(ProjectHeader);
        expect(headers.length).toBe(2);
      });
    });

    describe('environment component', () => {
      it('should have one environment component per environment', () => {
        const environments = wrapper.findAll(Environment);
        expect(environments.length).toBe(3);
      });
    });
  });
});
