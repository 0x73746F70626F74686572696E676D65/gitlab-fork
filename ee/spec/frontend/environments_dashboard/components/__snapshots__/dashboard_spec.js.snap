// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard should match the snapshot 1`] = `
<div
  class="operations-dashboard"
>
  <div>
    <!---->
  </div>
   
  <div
    class="page-title-holder flex-fill d-flex align-items-center"
  >
    <h1
      class="js-dashboard-title page-title text-nowrap flex-fill"
    >
      
      Environments Dashboard
    
    </h1>
     
    <button
      class="btn js-add-projects-button btn btn-success btn-secondary"
      type="button"
    >
      
      Add projects
    
    </button>
  </div>
   
  <div
    class="prepend-top-default"
  >
    <div
      class="row prepend-top-20 text-center"
    >
      <div
        class="col-12 d-flex justify-content-center svg-content"
      >
        <img
          class="js-empty-state-svg col-12 prepend-top-20"
          src="/assets/illustrations/operations-dashboard_empty.svg"
        />
      </div>
       
      <h4
        class="js-title col-12 prepend-top-20"
      >
        
        Add a project to the dashboard
      
      </h4>
       
      <div
        class="col-12 d-flex justify-content-center"
      >
        <span
          class="js-sub-title mw-460 text-tertiary text-left"
        >
          
          The environments dashboard provides a summary of each project's environments' status, including pipeline and alert statuses.
        
        </span>
      </div>
       
      <div
        class="col-12"
      >
        <a
          class="js-documentation-link btn btn-primary prepend-top-default append-bottom-default"
          href="/help/user/operations_dashboard/index.html"
        >
          
          View documentation
        
        </a>
      </div>
    </div>
  </div>
</div>
`;
